torch/cuda/random.py

```diff
@@ -1,26 +1,55 @@
 from torch import _C
-from . import _lazy_init, _lazy_call
+from . import _lazy_init, _lazy_call, device_count, device
 
 
-def get_rng_state():
+def get_rng_state(device_id=-1):
     r"""Returns the random number generator state of the current
     GPU as a ByteTensor.
 
+    Args:
+        device_id (int, default -1): The device to return the RNG state of;
+            by default we use the current device
+
     .. warning::
         This function eagerly initializes CUDA.
     """
     _lazy_init()
-    return _C._cuda_getRNGState()
+    with device(device_id):
+        return _C._cuda_getRNGState()
+
 
+def get_rng_state_all():
+    r"""Returns a tuple of ByteTensor representing the random number states of all devices."""
 
-def set_rng_state(new_state):
+    results = []
+    for i in range(device_count()):
+        with device(i):
+            results.append(get_rng_state())
+    return results
+
+
+def set_rng_state(new_state, device_id=-1):
     r"""Sets the random number generator state of the current GPU.
 
     Args:
         new_state (torch.ByteTensor): The desired state
     """
     new_state_copy = new_state.clone()
-    _lazy_call(lambda: _C._cuda_setRNGState(new_state_copy))
+
+    def cb():
+        with device(device_id):
```

Comment:
You should probably set `device_id` to a concrete idx before you capture it in this lambda, because it might be different at the time lazy calls are processed

Reply:
It is subtle (I will add a comment), but actually, you can't do this. To determine the actual "current" device id requires you to initialize the CUDA driver, which defeats the point of laziness. Fortunately, if `_lazy_call` doesn't execute the `setRNGState` immediately, it *must* be the case that the current device is the default, because there's no way to set the device without initializing CUDA.